### ISSUE TITLE
VIM editior in Docker container in GNS3 not working

### ISSUE THREAD

**Issue by ehlers** [GNS3/gns3-gui#2136]
Issue in the forum: https://www.gns3.com/qa/vim-editior-in-docker-container-

I can reproduce that with this tiny Dockerfile:

```
FROM debian:stretch-slim
RUN apt-get update && apt-get -y install vim
```

When starting vim in the container, the console closes and the container stops. That doesn't happen, when I start the container manually with `docker run -ti`. Workaround is to change the TERM environment variable from the default value xterm to vt100.

When I start gns3server with debug, the only output after I started vim is:
```
2017-07-02 11:04:24 INFO hypervisor.py:186 Stopping uBridge process PID=6267
2017-07-02 11:04:30 INFO docker_vm.py:557 Docker container 'vim-1' [vim:latest] stopped
```

**Comment by ehlers** [comment]
One observation during a complete unrelated docker session:

When the docker VM sends an invalid UTF-8 sequence, the console terminal closes and the VM stop, exactly as in this VIM issue. Doing that in a "docker run" has no bad side effect. Maybe VIM issues a character sequence when using TERM=xterm that is invalid UTF-8???

To send an invalid UTF-8 to the console:
```
/bin/echo -e '\xc3\x23'
```

**Comment by ehlers** [comment]
Verified it. In a script session started "docker run -ti vim" and then started vim. The script output shows an invalid UTF-8 character hex "bd" at offset b9.

```
00000000  53 63 72 69 70 74 20 73  74 61 72 74 65 64 20 6f  |Script started o|
00000010  6e 20 54 75 65 20 30 34  20 4a 75 6c 20 32 30 31  |n Tue 04 Jul 201|
00000020  37 20 31 32 3a 30 31 3a  31 34 20 41 4d 20 43 45  |7 12:01:14 AM CE|
00000030  53 54 0a 1b 5d 30 3b 67  6e 73 33 40 67 6e 73 33  |ST..]0;gns3@gns3|
00000040  76 6d 3a 20 7e 07 67 6e  73 33 40 67 6e 73 33 76  |vm: ~.gns3@gns3v|
00000050  6d 3a 7e 24 20 64 6f 63  6b 65 72 20 72 75 6e 20  |m:~$ docker run |
00000060  2d 74 69 20 76 69 6d 0d  0a 72 6f 6f 74 40 33 36  |-ti vim..root@36|
00000070  65 35 30 31 65 63 32 61  37 65 3a 2f 23 20 0d 1b  |e501ec2a7e:/# ..|
00000080  5b 4b 72 6f 6f 74 40 33  36 65 35 30 31 65 63 32  |[Kroot@36e501ec2|
00000090  61 37 65 3a 2f 23 20 76  69 6d 0d 0a 1b 5b 3f 31  |a7e:/# vim...[?1|
000000a0  30 30 30 68 1b 5b 3f 31  30 34 39 68 1b 5b 3f 31  |000h.[?1049h.[?1|
000000b0  68 1b 3d 1b 5b 32 3b 31  48 bd 1b 5b 36 6e 1b 5b  |h.=.[2;1H..[6n.[|
000000c0  32 3b 31 48 20 20 1b 5b  31 3b 31 48 1b 5d 31 31  |2;1H  .[1;1H.]11|
000000d0  3b 3f 07 1b 5b 31 3b 32  35 72 1b 5b 3f 31 32 3b  |;?..[1;25r.[?12;|
000000e0  32 35 68 1b 5b 3f 31 32  6c 1b 5b 3f 32 35 68 1b  |25h.[?12l.[?25h.|
.
.
```

Copying this script output to a docker container running within GNS3 and sending that to the console (cat typescript) again closes the console and stops the docker VM.

So I have the strong assumption, that invalid UTF-8 creates this issue. Hopefully it's possible to transport the data without interpretation.

**Comment by ziajka** [comment]
Docker CE 17.06 comes with binary mode for WebSocket which allows to send invalid UTF-8 data, reference: https://github.com/moby/moby/pull/30460. It is needed to adjust GNS3 to support binary mode instead of text mode.

**Comment by ehlers** [comment]
But that leaves only 17.06 as the only stable docker version, that works with GNS3. Then almost all GNS3 installation have to be updated.

Personally I would rather live with that bug until that (or a later) docker version is more widely deployed.

**Comment by noplay** [comment]
I think we can't keep this bug it's too easy to trigger if you send random data to the terminal.

Using the binary websocket seem to be correct solution, all other solution seem hacky. 

For docker  I'm for ignoring what distribution support because it's not the recommended way to install docker. It's a software that move fast and not yet stable. Especially the package from debian/ubuntu is known to have been buggy I would not trust it. 

But we can include the patch in GNS3 2.1 to avoid to introduce such change in a patch release.

**Comment by ziajka** [comment]
Proposed solution works in a way we're backward compatible and when user has newer Docker version then we take an advantage of that.

**Comment by luizsantos** [comment]
Hi everyone, this problem was fixed? because I'm using gns3 version 2.2.45 and I have the same problem. If I execute vi command, GNS3 stop/exit the docker container. I executed also  /bin/echo -e '\xc3\x23' and the gns3 docker console exit too. Someone know how to fix this?

**Comment by luizsantos** [comment]
Well, I used vim-tiny instead vim and it's worked...

> Hi everyone, this problem was fixed? because I'm using gns3 version 2.2.45 and I have the same problem. If I execute vi command, GNS3 stop/exit the docker container. I executed also /bin/echo -e '\xc3\x23' and the gns3 docker console exit too. Someone know how to fix this?

**Comment by grossmj** [comment]
Thanks for the feedback 👍
